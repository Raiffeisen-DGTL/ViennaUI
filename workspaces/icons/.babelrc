{
  "presets": ["@babel/env", "@babel/typescript", "@babel/react"],
  "env": {
    "esm": {
      "presets": [["@babel/env", { "modules": false }]]
    }
  }
}
