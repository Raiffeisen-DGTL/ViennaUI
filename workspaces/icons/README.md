# Библиотека иконок

###### Raiffeisenbank Design System

Коллекция иконок для использования в интерфейсах. Иконки представлены в виде React компонентов и оригинальных векторных (`*.svg`) изображений. Изображения напрямую можно использовать из пакета.
