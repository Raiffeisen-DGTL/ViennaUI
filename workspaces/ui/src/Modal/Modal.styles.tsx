import styled, { css } from 'styled-components';
import { getPresets } from 'vienna.ui-primitives';

const fade = getPresets('modal.fade', {
    base: null,
    hide: null,
    show: null,
    custom: null,
});

const modal = getPresets('modal.window', {
    base: null,
    hide: null,
    show: null,
    custom: null,
});

const icon = getPresets('modal.icon', {
    base: null,
    hover: null,
    custom: null,
});

export const Fade = styled.div<{ show?: boolean }>`
    position: fixed;
    display: flex;
    align-items: center;
    flex-direction: column;
    overflow-y: auto;
    top: 0;
    left: 0;
    right: 0;
    bottom: 0;

    ${fade.base}
    ${fade.custom}
    ${fade.hide}

    visibility: hidden;

    ${({ show }) =>
        show &&
        css`
            visibility: visible;
            ${fade.show}
        `}
`;

export const Box = styled.div<{ toggle?: boolean }>`
    ${modal.base}
    -webkit-font-smoothing: antialiased;
    position: relative;
    margin: auto;
    ${modal.hide}
    ${modal.custom}
    ${({ toggle }) =>
        toggle &&
        css`
            ${modal.show}
        `};
`;

export const CloseIcon = styled.div`
    ${icon.base}
    &:hover {
        ${icon.hover}
    }
    ${icon.custom}
`;
