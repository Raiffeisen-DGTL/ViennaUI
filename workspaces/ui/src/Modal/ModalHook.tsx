import React, { useState } from 'react';
import ReactDOM from 'react-dom';
import { Modal } from './Modal';

let modals: React.ReactNode[] = [];
let setModals: React.Dispatch<React.ReactNode[]>;
let container: HTMLDivElement;

const ModalContainer = () => {
    const [modals, _setModals] = useState<React.ReactNode[]>([]);
    setModals = _setModals;
    return <>{modals.map((modal) => modal)}</>;
};

type ModalHookOpen = () => void;
type ModalHookClose = <T extends any>(data?: T) => void;

interface ModalHook {
    open: ModalHookOpen;
    close: ModalHookClose;
}

export function useModal(): ModalHook;
export function useModal(modal: React.ReactNode, onClose: ModalHookClose): [ModalHookOpen, ModalHookClose];
export function useModal(modal?: any, onClose?: any): any {
    const modalRef: any = { current: { close: null } };
    let instanse: React.ReactNode;

    const closeHandler: ModalHookClose = (data) => {
        modals = modals.filter((m) => m !== instanse);
        if (typeof onClose === 'function') {
            onClose(data);
        }
    };

    const close: ModalHookClose = (data) => {
        if (typeof modalRef.current.close === 'function' && modals.includes(instanse)) {
            modalRef.current.close(data);
        }
    };

    const open: ModalHookOpen = () => {
        if (!container) {
            container = document.createElement('div');
            document.body.appendChild(container);
            ReactDOM.render(<ModalContainer />, container);
        }

        const state: any = { open: null };

        const handleRef = (ref) => {
            modalRef.current = ref;
            state.open();
        };

        instanse = (
            <Modal key={Math.random()} state={state} ref={handleRef} onClose={closeHandler}>
                {modal}
            </Modal>
        );

        modals.push(instanse);
        if (setModals) {
            setModals([...modals]);
        }
    };

    if (!modal && !onClose) {
        return {
            set children(val) {
                modal = val;
            },
            set onClose(val) {
                onClose = val;
            },
            set open(val) {
                this.nativeOpen = val;
            },
            get open() {
                return open;
            },
            set close(val) {
                this.nativeClose = val;
            },
            get close() {
                return close;
            },
        };
    }

    return [open, close];
}
