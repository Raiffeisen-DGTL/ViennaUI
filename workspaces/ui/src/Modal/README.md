# Modal

Используется в случаях:

-   когда необходимо перевести внимание пользователя от текущей задачи к чему-то более важному
-   сообщить об ошибках, положительном завершении работы над задачей, уточнить намерения пользователя
-   показать дополнительную информацию/действие внутри контекста

## Импорт

```
import { Modal, useModal } from 'vienna-ui';
```

## Свойства / Props

Prop | Type | Default | Description
--- | --- | --- | ---
isOpen | boolean \| undefined | false |
closeIcon | ReactNode | <Close /> | Иконка для закрытия
onClose | ((data?: any) => boolean \| void \| Promise) \| undefined | false |
onPreDispose | (() => void) \| undefined | false | Обработчик события исчезновения модального окна
state | any | false | Управление состоянием черех хук `useModal`

## Использование

Компонент состоит из родительского контейнера `Modal` и элементов контента `Modal.Layout`, `Modal.Head`, `Modal.Title`, `Modal.SubTitle`, `Modal.Body` и `Modal.Footer`.

```jsx
const state = useModal(); //при использовании в декларативном стиле

// state.open() - открыть окно
// state.close(data) - закрыть окно с передачей в onClose - data

closeHandler = (data) => {
    /**/
};

// ...

<Modal state={state} onClose={closeHandler}>
    content
</Modal>;

// ...
```

или

```jsx
const isOpen = useState(false);

<Modal isOpen={isOpen} onClose={closeHandler}>
    content
</Modal>;
```

##### Модальное окно с отступами: заголовок, контент, подвал. (пример без хуков)

Является самым предпочтительным вариантом для большинства случаев. Особенно если требуется обновлять состояние содержимого модального окна из родителя.

```jsx
{() => {
    const [isOpen, setIsOpen] = React.useState(false);
    return (
        <>
            <Button onClick={() => setIsOpen(true)}>Show modal</Button>
            <Modal isOpen={isOpen} onClose={() => setIsOpen(false)}>
                <Modal.Layout>
                    <Modal.Head>
                        <Modal.Title>Modal header</Modal.Title>
                    </Modal.Head>
                    <Modal.Body>
                        <Demo>
                            Lorem ipsum dolor sit amet, consectetur adipiscing elit, sed do eiusmod tempor
                            incididunt ut labore et dolore magna aliqua. Ut enim ad minim veniam, quis nostrud
                            exercitation ullamco laboris.
                        </Demo>
                    </Modal.Body>
                    <Modal.Footer>
                        <Groups>
                            <Button size='l' design='outline' onClick={() => setIsOpen(false)}>
                                Button example
                            </Button>
                            <Button size='l' design='accent' onClick={() => setIsOpen(false)}>
                                Button example
                            </Button>
                        </Groups>
                    </Modal.Footer>
                </Modal.Layout>
            </Modal>
        </>
    );
}}
```

## Использование в режиме менеджера
##### Свойство `state`, `useModal`

```
const [open, close] = useModal(ModalContent, callback); // при использовании в режиме менеджера

// open() - открыть окно
// close(data) - закрыть окно с передачей в callback - data

// ...

open();
// close(data)

// ...
```

> Данный способ дает возможность создавать переиспользуемые модальные окна, которые можно открывать из любой части приложения не привязываясь к контенту или компонентам.
> При передаче в `useModal` в качестве параметров `JSX.Elemen`'а контента окна и `callback` функции срабатывающей на `onClose`, возвращаются методы открытия и закрытия окна в режиме менеджера.

```jsx
{() => {
    const [open, close] = useModal(
        <Modal.Layout>
            <Image />
            <Modal.Head align={'center'}>
                <Modal.Title>Modal header</Modal.Title>
            </Modal.Head>
            <Modal.Body>
                <Demo style={{ textAlign: 'center' }}>
                    Lorem ipsum dolor sit amet, consectetur adipiscing elit, sed do eiusmod tempor incididunt ut
                    labore et dolore magna aliqua.
                </Demo>
            </Modal.Body>
            <Modal.Footer align={'center'}>
                <Button size='l' design='accent'>
                    Button example
                </Button>
            </Modal.Footer>
        </Modal.Layout>
    );
    return <Button onClick={open}>Show modal</Button>;
}}
```

## Признак открытого компонента
##### Свойство `isOpen`

Для того, чтобы управлять состояние открытия/закрытия модального окна необходимо передать соотвествующий статус `isOpen`.

```
<Button onClick={() => setIsOpen(true)}>Show modal</Button>
<Modal isOpen={isOpen} onClose={() => setIsOpen(false)}>
    <Modal.Layout>
        <Modal.Head>
            <Modal.Title>Modal header</Modal.Title>
        </Modal.Head>
        <Modal.Body>
            <Demo>
                Lorem ipsum dolor sit amet, consectetur adipiscing elit, sed do eiusmod tempor
                incididunt ut labore et dolore magna aliqua. Ut enim ad minim veniam, quis nostrud
                exercitation ullamco laboris.
            </Demo>
        </Modal.Body>
        <Modal.Footer>
            <Groups>
                <Button size='l' design='outline' onClick={() => setIsOpen(false)}>
                    Button example
                </Button>
                <Button size='l' design='accent' onClick={() => setIsOpen(false)}>
                    Button example
                </Button>
            </Groups>
        </Modal.Footer>
    </Modal.Layout>
</Modal> 
```

## Иконка для закрытия
##### Свойство `closeIcon`

В качестве иконки для закрытия компонента, можно передать любой компонент.

```
<Button onClick={() => setIsOpen(true)}>Show modal</Button>
<Modal closeIcon={<>Close</>} isOpen={isOpen} onClose={() => setIsOpen(false)}>
    <Modal.Layout>
        <Modal.Head>
            <Modal.Title>Modal header</Modal.Title>
        </Modal.Head>
        <Modal.Body>
            <Demo>
                Lorem ipsum dolor sit amet, consectetur adipiscing elit, sed do eiusmod tempor
                incididunt ut labore et dolore magna aliqua. Ut enim ad minim veniam, quis nostrud
                exercitation ullamco laboris.
            </Demo>
        </Modal.Body>
        <Modal.Footer>
            <Groups>
                <Button size='l' design='outline' onClick={() => setIsOpen(false)}>
                    Button example
                </Button>
                <Button size='l' design='accent' onClick={() => setIsOpen(false)}>
                    Button example
                </Button>
            </Groups>
        </Modal.Footer>
    </Modal.Layout>
</Modal> 
```

## Исчезновение компонента
##### Свойство `onPreDispose`

Метод `onPreDispose` вызывается после завершения анимации закрытия, непосредственно перед удалением модального окна из `DOM`.

```
<Modal
    isOpen={isOpen}
    onClose={handleCustomClose}
    onPreDespose={() => console.log('end despose')}
>
    <Modal.Layout>
        <Modal.Head>
            <Modal.Title>Modal header</Modal.Title>
        </Modal.Head>
        <Modal.Body>
            <Demo>
                Lorem ipsum dolor sit amet, consectetur adipiscing elit, sed do eiusmod tempor
                incididunt ut labore et dolore magna aliqua. Ut enim ad minim veniam, quis nostrud
                exercitation ullamco laboris.
            </Demo>
        </Modal.Body>
        <Modal.Footer>
            <Groups>
                <Button size='l' design='outline' onClick={() => setIsOpen(false)}>
                    Button example
                </Button>
                <Button size='l' design='accent' onClick={() => setIsOpen(false)}>
                    Button example
                </Button>
            </Groups>
        </Modal.Footer>
    </Modal.Layout>
</Modal>
```
