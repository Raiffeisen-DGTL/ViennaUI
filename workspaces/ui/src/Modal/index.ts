export * from './Modal';
export * from './ModalHook';
