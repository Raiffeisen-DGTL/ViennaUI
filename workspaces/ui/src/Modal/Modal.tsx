import React, { useCallback, useState, useEffect, ForwardRefExoticComponent, RefAttributes, useRef } from 'react';
import ReactDOM from 'react-dom';
import { useDebounce } from 'vienna.react-use';
import { Close } from 'vienna.icons';
import { Screen, ScreenSlots } from 'vienna.ui-primitives';
import { Fade, Box, CloseIcon } from './Modal.styles';

const ANIMATION_DURATION = 400;

export interface ModalProps extends React.PropsWithChildren<any> {
    id?: string;
    state?: any;
    isOpen?: boolean;
    closeIcon?: React.ReactNode;
    onClose?: (data?: any) => void | boolean | Promise<boolean>;
    onPreDispose?: () => void;
}

export interface ModalSlots extends ScreenSlots {
    Layout: typeof Screen;
}

export const Modal: React.FC<ModalProps> & ModalSlots = React.forwardRef((props: ModalProps, ref: any) => {
    const { id, children, state, onClose, onPreDispose, isOpen, closeIcon = <Close size='l' /> } = props;

    const [init, setInit] = useState(false);
    const [show, setShow] = useState(false);
    const [toggle, setToggle] = useState(false);

    if (state && isOpen) {
        // eslint-disable-next-line no-console
        console.warn('Нельзя одновременно использовать свойства state и isOpen');
    }

    // need for styled-components accept the changes
    const initDebouncer = useDebounce(setInit, ANIMATION_DURATION);
    const toggleDebouncer = useDebounce(setToggle, 100);

    const predesposeHandler = useCallback(() => {
        if (typeof onPreDispose === 'function') {
            onPreDispose();
        }
    }, [onPreDispose]);

    // close animation process
    const playCloseAnimation = useCallback(() => {
        // hide modal box
        toggleDebouncer(false, () => {
            // hide fade
            setShow(false);
        });
        // unmount
        initDebouncer(false, predesposeHandler);
    }, [toggleDebouncer, initDebouncer, predesposeHandler]);

    const close = useCallback(
        (data, force = false) => {
            if (force || typeof onClose !== 'function') {
                playCloseAnimation();
                return;
            }
            const flag = onClose(data);
            if (flag instanceof Promise) {
                flag.then(
                    (flag) => {
                        if (flag || flag === undefined) {
                            playCloseAnimation();
                        }
                    },
                    // eslint-disable-next-line no-console
                    (error) => console.error(error)
                );
                return;
            }
            if (flag === true || flag === undefined) {
                playCloseAnimation();
            }
        },
        [onClose, playCloseAnimation]
    );

    // add methods to incoming modalHook
    if (state) {
        state.children = children;
        state.onClose = onClose;
        state.open = () => setInit(true);
        state.close = close;
    }

    const prevTarget = useRef();
    const fadeMouseDown = useCallback(
        (e) => {
            prevTarget.current = e.target;
        },
        [prevTarget]
    );

    const handleClickFade = useCallback(
        (e) => {
            // Ignore the events not coming from the "fade"
            // We don't want to close the dialog when clicking the dialog content.
            if (e.target !== e.currentTarget) {
                return;
            }

            // Make sure the event starts and ends on the same DOM element.
            if (prevTarget.current !== e.target) {
                return;
            }
            close(null);
        },
        [close]
    );

    const handleClose = useCallback(() => close(null), [close]);

    const boxClickHandler = useCallback((event) => event.stopPropagation(), []);

    const boxMouseDown = useCallback((event) => event.stopPropagation(), []);

    // ***** ANIMATION START

    // when isOpen set to true - create portal !!! DONT ADD init TO DEPS ARRAY !!!
    useEffect(() => {
        if (typeof isOpen === 'boolean') {
            if (isOpen) {
                setInit(true);
            } else if (init) {
                close(null, true);
            }
        }
    }, [isOpen, close]);

    // when portal created show fade
    useEffect(() => {
        if (init) {
            // show fade
            setShow(true);
            // show modal - need foe correct animation
            toggleDebouncer(true);
        }
    }, [init, toggleDebouncer]);

    // ***** ANIMATION END

    useEffect(() => {
        if (ref) {
            ref({ close: (data?) => close(data) });
        }
    }, [ref, close]);

    // if we unmount component - despose debouncers
    useEffect(
        () => () => {
            initDebouncer.despose();
            toggleDebouncer.despose();
        },
        []
    );

    if (typeof window === 'undefined') {
        return null;
    }

    return (
        document &&
        ReactDOM.createPortal(
            init && (
                <Fade id={id} show={show} onClick={handleClickFade} onMouseDown={fadeMouseDown}>
                    <Box toggle={toggle} onClick={boxClickHandler} onMouseDown={boxMouseDown}>
                        {children}
                        {closeIcon && <CloseIcon onClick={handleClose}>{closeIcon}</CloseIcon>}
                    </Box>
                </Fade>
            ),
            document.body
        )
    );
}) as ForwardRefExoticComponent<ModalProps & RefAttributes<any>> & ModalSlots;

Modal.displayName = 'Modal';

Modal.Layout = Screen;
Modal.Head = Screen.Head;
Modal.Title = Screen.Title;
Modal.SubTitle = Screen.SubTitle;
Modal.Body = Screen.Body;
Modal.Footer = Screen.Footer;
export default Modal;
