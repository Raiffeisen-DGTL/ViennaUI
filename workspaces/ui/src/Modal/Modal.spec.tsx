import React from 'react';
import { Close } from 'vienna.icons';
import { Fade, Box, CloseIcon } from './Modal.styles';

test('Modal', () => {
    const snap = snapshot.render(
        <Fade>
            <Box>
                <CloseIcon>
                    <Close />
                </CloseIcon>
            </Box>
        </Fade>
    );
    expect(snap).toMatchSnapshot();
});

test('Modal /w show', () => {
    const snap = snapshot.render(
        <Fade show>
            <Box toggle>
                <CloseIcon>
                    <Close />
                </CloseIcon>
            </Box>
        </Fade>
    );
    expect(snap).toMatchSnapshot();
});
