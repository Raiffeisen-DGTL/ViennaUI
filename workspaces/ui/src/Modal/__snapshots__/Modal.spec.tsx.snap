// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Modal /w show 1`] = `
.c0 {
  position: fixed;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  overflow-y: auto;
  top: 0;
  left: 0;
  right: 0;
  bottom: 0;
  z-index: 1000;
  -webkit-transition: all 0.3s ease-out;
  transition: all 0.3s ease-out;
  padding: 32px;
  background-color: rgba(43,45,51,0);
  visibility: hidden;
  visibility: visible;
  background-color: rgba(43,45,51,0.24);
}

.c1 {
  z-index: 2;
  -webkit-transition: all 0.3s ease-in-out;
  transition: all 0.3s ease-in-out;
  box-shadow: 0 8px 32px 0 rgba(0,0,0,0.08);
  border-radius: 16px;
  background-color: #FFFFFF;
  opacity: 0;
  -webkit-font-smoothing: antialiased;
  position: relative;
  margin: auto;
  -webkit-transform: scale(0.8);
  -ms-transform: scale(0.8);
  transform: scale(0.8);
  opacity: 0;
  opacity: 1;
  -webkit-transform: scale(1);
  -ms-transform: scale(1);
  transform: scale(1);
}

.c2 {
  color: #AAABAD;
  position: absolute;
  right: 16px;
  top: 16px;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  background-color: rgba(255,255,255,0.64);
  width: 32px;
  height: 32px;
  border-radius: 16px;
  cursor: pointer;
  -webkit-transition: all .2s;
  transition: all .2s;
  outline: none;
}

.c2:hover {
  color: #2B2D33;
}

<div
  class="c0"
>
  <div
    class="c1"
  >
    <div
      class="c2"
    >
      <svg
        fill="currentColor"
        height="20"
        viewBox="0 0 24 24"
        width="20"
        xmlns="http://www.w3.org/2000/svg"
      >
        <defs />
        <path
          d="M12 13.414l-7.293 7.293-1.414-1.414L10.586 12 3.293 4.707l1.414-1.414L12 10.586l7.293-7.293 1.414 1.414L13.414 12l7.293 7.293-1.414 1.414L12 13.414z"
        />
      </svg>
    </div>
  </div>
</div>
`;

exports[`Modal 1`] = `
.c0 {
  position: fixed;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  overflow-y: auto;
  top: 0;
  left: 0;
  right: 0;
  bottom: 0;
  z-index: 1000;
  -webkit-transition: all 0.3s ease-out;
  transition: all 0.3s ease-out;
  padding: 32px;
  background-color: rgba(43,45,51,0);
  visibility: hidden;
}

.c1 {
  z-index: 2;
  -webkit-transition: all 0.3s ease-in-out;
  transition: all 0.3s ease-in-out;
  box-shadow: 0 8px 32px 0 rgba(0,0,0,0.08);
  border-radius: 16px;
  background-color: #FFFFFF;
  opacity: 0;
  -webkit-font-smoothing: antialiased;
  position: relative;
  margin: auto;
  -webkit-transform: scale(0.8);
  -ms-transform: scale(0.8);
  transform: scale(0.8);
  opacity: 0;
}

.c2 {
  color: #AAABAD;
  position: absolute;
  right: 16px;
  top: 16px;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  background-color: rgba(255,255,255,0.64);
  width: 32px;
  height: 32px;
  border-radius: 16px;
  cursor: pointer;
  -webkit-transition: all .2s;
  transition: all .2s;
  outline: none;
}

.c2:hover {
  color: #2B2D33;
}

<div
  class="c0"
>
  <div
    class="c1"
  >
    <div
      class="c2"
    >
      <svg
        fill="currentColor"
        height="20"
        viewBox="0 0 24 24"
        width="20"
        xmlns="http://www.w3.org/2000/svg"
      >
        <defs />
        <path
          d="M12 13.414l-7.293 7.293-1.414-1.414L10.586 12 3.293 4.707l1.414-1.414L12 10.586l7.293-7.293 1.414 1.414L13.414 12l7.293 7.293-1.414 1.414L12 13.414z"
        />
      </svg>
    </div>
  </div>
</div>
`;
